lib/dynamoid/transaction/upsert.rb

```diff
@@ -0,0 +1,32 @@
+require_relative 'action'
+
+module Dynamoid
+  class Transaction
+    class Upsert < Action
+
+      def run_callbacks
+        model.run_callbacks(:save) do
+          # We don't know if this is a new record or not so skip :create and :update callbacks
+          model.run_callbacks(:validate) do
+            yield if block_given?
+          end
+        end
+      end
+
+      def to_h
+        changes = model.changes.map{|k,v| [k.to_sym,v[1]]}.to_h # hash of dirty attributes
+        changes = changes.merge attributes if attributes.present?
+        changes = add_timestamps(changes, skip_created_at: false) # overwrites created_at since this will be a new record essentially
+        changes[model_class.hash_key] ||= model.hash_key || SecureRandom.uuid
+        changes[model_class.range_key] ||= model.range_key if model_class.range_key? && model.range_key.present?
+        item = Dynamoid::Dumping.dump_attributes(changes, model_class.attributes)
+        {
+          put: {
+            item: sanitize_item(item),
+            table_name: model_class.table_name,
+          }
```

Comment:
It seems `put` that behaves like `PutItem` operation and just **replaces** existing item with provided attributes when item with primary key exists. But we need upserting - **partial updating** of existing item. So I suppose `update` operation should be used.